Detect external modification of scenes

KoBeWi opened the pull request:
With this change, editor will look for external modifications in opened scenes when focused and asks for reload. This is analogical to scripts in ScriptEditorPlugin.

Closes #11803
Also closes #20250, which is a duplicate
Also closes #22450, which is a duplicate duplicate

KoBeWi committed:
Detect external modification of scenes

KoBeWi commented:
I finally got around to fixing this and the crash is gone (not sure if related to what I did xd). Also the reloading is more seamless, there's no undo history noise at all.

Probably needs some testing, but otherwise, the PR complete. I'd really like it for 3.2 >_>

akien-mga commented:
We're a bit too close to 3.2-stable for such a change IMO, so I'd prefer to merge after the release. We can cherry-pick for 3.2 though if it proves to work well in the master branch.

alexzheng commented:
Why such a nice feature still not available in the latest release?
Scene can not be reload automatically will overwrite the external modification.

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -6483,6 +6556,28 @@ EditorNode::EditorNode() {
+		disk_changed->add_child(vbc);
+
+		Label *dl = memnew(Label);
+		dl->set_text(TTR("The following files are newer on disk.\nWhat action should be taken?:"));
```
`?:` is invalid punctuation, remove `:`.

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -592,6 +593,72 @@ void EditorNode::_sources_changed(bool p_exist) {
+	disk_changed->hide();
+}
+
+void EditorNode::_reload_scenes() {
```
I'd name it `reload_modified_scenes()` since it doesn't reload non-modified scenes a priori.

KoBeWi commented:
Ok, rebased and fixed. However due to #37531, the dialog will pop up infinitely when you try to cancel/close it (because editor gets refocused), so you are either forced to save scene or reload.

Dragoncraft89 commented:
As far as I can tell from the diff, this only takes into account open scenes, right?

But these are not the only files, that need to be reloaded. The `project.godot` file faces similar problems, although fixing that could be a lot more difficult

Zylann commented:
Note that the user can have more than just scenes open. It's actually about *resources*. Resources can be scripts, scenes, or standalone resources open in the inspector. Is the latter handled as well?

KoBeWi commented:
Resources in Godot are usually built-in, so this PR handles most of the problems. But I didn't test file Resources, they might indeed be a problem. project.godot is not handled yet either.

Zylann commented:
> Resources in Godot are usually built-in

No? I dunno what that comes from (we dont have stats?) but there are a bunch of cases where resources are files. I also use lots of file resources myself, and actually rarely make them builtin.

KoBeWi commented:
Dunno, 99% of non-imported resources in my project are built-in (they are mostly collision shapes). Script changes are already handled by script editor.

KoBeWi committed:
Detect external modification of project.godot

KoBeWi commented:
I rebased the PR. Also I added detecting changes to `project.godot`, but right now it's useless due to #45243 (I tested with the mentioned lines commented out and it works)

KoBeWi commented:
Ah, I forgot to write here, but I tested this PR during the GGJ and it works fine.

There are two problems though:
- #37531
- built-in scripts aren't properly reloaded when the containing scene is instanced in another opened scene

However they are separate issues not really related to this PR. I'd say it's ready to be merged.

akien-mga commented on core/config/project_settings.cpp:
```diff
@@ -676,7 +677,12 @@ void ProjectSettings::clear(const String &p_name) {
 
 Error ProjectSettings::save() {
-	return save_custom(get_resource_path().plus_file("project.godot"));
+	print_line("save");
```
That's going to get annoying pretty fast :)

KoBeWi replied:
Oops

akien-mga commented on core/config/project_settings.h:
```diff
@@ -113,7 +115,6 @@ class ProjectSettings : public Object {
 
 	Error _setup(const String &p_path, const String &p_main_pack, bool p_upwards = false);
 
-protected:
```
Why? `_bind_methods()` is always `protected`.

KoBeWi replied:
It's already in the `protected` block, scroll up a bit.

akien-mga commented on editor/editor_data.h:
```diff
@@ -111,6 +111,7 @@ class EditorData {
 	struct EditedScene {
 		Node *root = nullptr;
 		String path;
+		uint64_t file_modified_time;
```
```suggestion
		uint64_t file_modified_time = 0;
```

Calinou commented on editor/editor_node.cpp:
```diff
@@ -880,6 +881,81 @@ void EditorNode::_sources_changed(bool p_exist) {
+
+			Error err = load_scene(filename, false, false, true, false, true);
+			if (err != OK) {
+				ERR_PRINT("Failed to load scene");
```
Make the error message give some context for easier debugging:

```suggestion
				ERR_PRINT(vformat("Failed to load scene: %s", filename));
```

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.2.4. (Tentatively to get it well tested in the next RCs - if there are regressions I might revert and re-cherry-pick for 3.2.5 instead.)

akien-mga commented on editor/editor_node.cpp:
```diff
@@ -6599,6 +6678,30 @@ EditorNode::EditorNode() {
+		vbc->add_child(disk_changed_list);
+		disk_changed_list->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+
+		disk_changed->connect("confirmed", callable_mp(this, &EditorNode::_reload_modified_scenes));
+		disk_changed->connect("confirmed", callable_mp(this, &EditorNode::_reload_project_settings));
```
You're connecting two callables to the same signal, bug?

KoBeWi replied:
No, this is intended. I made reloading scenes and project settings into separate methods.

akien-mga commented:
If you're up for it, I'd suggest making a dedicated backport PR for `3.2`, as the cherry-picking needs some care and proper testing to make sure that going back to `_bind_methods` for callbacks, etc. works as intended.
